sound/soc/sof/ipc.c

```diff
@@ -623,38 +797,25 @@ int snd_sof_ipc_set_comp_data(struct snd_sof_ipc *ipc,
 {
 	struct snd_sof_dev *sdev = ipc->sdev;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
-	int err;
+	int err = 0;
 
 	/* read firmware volume */
-	if (scontrol->readback_offset != 0) {
+	if (scontrol->readback_offset != 0)
 		/* we can read value header via mmaped region */
 		snd_sof_dsp_block_write(sdev, sdev->mmio_bar,
 					scontrol->readback_offset, cdata->chanv,
 					sizeof(struct sof_ipc_ctrl_value_chan) *
 					cdata->num_elems);
 
-	} else {
-		/* write value via slower IPC */
-		cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
-		cdata->cmd = ctrl_cmd;
-		cdata->type = ctrl_type;
-		cdata->rhdr.hdr.size = scontrol->size;
-		cdata->comp_id = scontrol->comp_id;
-		cdata->num_elems = scontrol->num_channels;
-
-		/* send IPC to the DSP */
-		err = sof_ipc_tx_message(sdev->ipc,
-					 cdata->rhdr.hdr.cmd, cdata,
-					 cdata->rhdr.hdr.size,
-					 cdata, cdata->rhdr.hdr.size);
-		if (err < 0) {
-			dev_err(sdev->dev, "error: failed to set control %d values\n",
-				cdata->comp_id);
-			return err;
-		}
-	}
+	else
+		err = sof_send_recv_ctrl_data(sdev, scontrol, ipc_cmd,
+					      ctrl_type, ctrl_cmd);
 
-	return 0;
+	if (err < 0)
+		dev_err(sdev->dev, "error: failed to set control %d values\n",
+			cdata->comp_id);
+
+	return err;
 }
 EXPORT_SYMBOL(snd_sof_ipc_set_comp_data);
```

Comment:
@juimonen why do we need separate snd_sof_ipc_set_comp_data and snd_sof_ipc_get_comp_data? They both do the same thing now isnt it?

Reply:
for that matter why do we need them at all. we should just call sof_send_recv_ctrl_data() from the caller

Reply:
well I was modifying the old code and not sure what was the idea... I'll see what I can do here.

Reply:
@ranj063 ok, modified also control.c and pm.c to call set_get_data directly.